text/0195-kubernetes-resource-detection.md

```diff
@@ -0,0 +1,113 @@
+
+As a cluster administrator, I can use my preferred policy agent or my own mutating admission webhook to add the OpenTelemetry-specified set of environment variables to all pods subsequently created in the cluster. Application developers deploying to the cluster just need to use the OpenTelemetry SDK with the Kubernetes resource detector.
+
+### User Journey: Kubernetes Config Management
+
+Kubernetes templating tools, such as [Helm templates](https://helm.sh/docs/chart_template_guide/), allow users to easily re-use snippets of yaml in many places in their configuration.  A common set of environment variables would users to define the OpenTelemetry env var config once, and use it across all of their OpenTelemetry deployments.
+
+**Note:** This pattern unfortunately isn't easy to use with kubernetes' built-in patching tools, [kustomize](https://kustomize.io/), because it doesn't [currently](https://github.com/kubernetes-sigs/kustomize/issues/1493) support patching all containers across deployments with a single patch.
+
+## Internal details
+
+Add an experimental k8s.md specification under specification/resource, which specifies how kubernetes resource detectors should detect the kubernetes semantic conventions:
+
+SDK's should add a Kubernetes environment variable-based resource detector.  The Kubernetes resource detector and the collector's `resourcedetectionprocessor` should support the following environment variables, and map them to the following semantic conventions:
+
+| Environment Variable | Semantic Convention |
+| ---- | ---- |
+| K8S_POD_NAME | k8s.pod.name |
+| K8S_POD_UID | k8s.pod.uid |
+| K8S_NAMESPACE_NAME | k8s.namespace.name |
+| K8S_NODE_NAME | k8s.node.name |
+
+Detection of the above semantic conventions using environment variables should not be done inside vendor-specific (e.g. GKE-specific) resource detectors.
+
+## Prior art and alternatives
+
+### Alternative: Detection required by SDKs
+
+Similar to `OTEL_RESOURCE_ATTRIBUTES`, we could require the proposed environment variables to always be detected.  While this would further improve the ease of use, [Resource semantic conventions](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/resource/semantic_conventions/README.md) (including Kubernetes) are still experimental, and the [SDK environment variable specification](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/sdk-environment-variables.md#general-sdk-configuration) are stable. Additionally, Kubernetes is [explicitly called out](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/resource/sdk.md#detecting-resource-information-from-the-environment) as a custom resource detector.
+
+### Alternative: Using HOSTNAME for k8s.pod.name
```

Comment:
I see some older issues ([e.g.](https://github.com/kubernetes/kubernetes/issues/4825)) that suggest that `HOSTNAME` is truncated to 64 bytes whereas the Pod name can be up to 253 bytes long. Should we add that as another argument against this alternative?

Reply:
added as a downside of this alternative.